File: 백준/Gold/2580. 스도쿠/스도쿠.py
sudoku = []
questions = []

for i in range(9):
    line = list(map(int, input().split()))
    sudoku.append(line)

    for j in range(9):
        if line[j] == 0:
            questions.append((i, j))

# def sol(x, y):
#     # 가로
#     for i in range(9):
#         if i == y:
#             continue
#
#         if sudoku[x][i] == sudoku[x][y]:
#             return False
#
#     # 세로
#     for i in range(9):
#         if i == x:
#             continue
#
#         if sudoku[i][y] == sudoku[x][y]:
#             return False
#
#     # 3 x 3
#     module_x = (x // 3) * 3
#     module_y = (y // 3) * 3
#
#     for i in range(module_x, module_x + 3):
#         for j in range(module_y, module_y + 3):
#             if i == x and j == y:
#                 continue
#
#             if sudoku[i][j] == sudoku[x][y]:
#                 return False
#
#     return True

# memo -> 실제 스도쿠 풀 때 칸에 후보군 메모하며 소거하던 거
row_memo = [set() for _ in range(9)]
column_memo = [set() for _ in range(9)]
module_memo = [set() for _ in range(9)]

# 초기 상태 설정
for i in range(9):
    for j in range(9):
        if sudoku[i][j] != 0:
            row_memo[i].add(sudoku[i][j])
            column_memo[j].add(sudoku[i][j])
            module_idx = (i // 3) * 3 + (j // 3)
            module_memo[module_idx].add(sudoku[i][j])

def backtrack(n):
    if n == len(questions):
        for line in sudoku:
            print(*line)

        exit()

    x, y = questions[n]
    module_idx = (x // 3) * 3 + (y // 3)
    for i in range(1, 10):

        if i not in row_memo[x] and i not in column_memo[y] and i not in module_memo[module_idx]:
            sudoku[x][y] = i
            row_memo[x].add(i)
            column_memo[y].add(i)
            module_memo[module_idx].add(i)

            if backtrack(n + 1):
                return True

            sudoku[x][y] = 0
            row_memo[x].remove(i)
            column_memo[y].remove(i)
            module_memo[module_idx].remove(i)

    return False

backtrack(0)

File: 백준/Gold/2580. 스도쿠/스도쿠.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 입력 처리
        int[][] sudoku = new int[9][9];
        List<int[]> blanks = new ArrayList<>();

        // 입력 받기
        for (int i = 0; i < 9; i++) {
            String line = br.readLine();
            String[] numbers = line.split(" ");

            for (int j = 0; j < 9; j++) {
                sudoku[i][j] = Integer.parseInt(numbers[j]);
                if (sudoku[i][j] == 0) {
                    blanks.add(new int[]{i, j});
                }
            }
        }

        // 풀이
        SudokuSolver solver = new SudokuSolver(sudoku, blanks);
        solver.solve();

        // 버퍼 정리
        br.close();
    }
}

class SudokuSolver {
    private final int[][] sudoku;
    private final List<int[]> blanks;
    private final List<Set<Integer>> rowMemo = new ArrayList<>(9);
    private final List<Set<Integer>> columnMemo = new ArrayList<>(9);
    private final List<Set<Integer>> moduleMemo = new ArrayList<>(9);

    public SudokuSolver(int[][] sudoku, List<int[]> blanks) {
        this.sudoku = sudoku;
        this.blanks = blanks;

        // Initialize lists with empty sets
        for (int i = 0; i < 9; i++) {
            this.rowMemo.add(new HashSet<>());
            this.columnMemo.add(new HashSet<>());
            this.moduleMemo.add(new HashSet<>());
        }

        this.initialize();
    }

    // 메모이제이션 초기화
    private void initialize() {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                int number = this.sudoku[i][j];

                if (number != 0) {
                    this.rowMemo.get(i).add(number);
                    this.columnMemo.get(j).add(number);

                    int moduleIndex = (i / 3) * 3 + (j / 3);
                    this.moduleMemo.get(moduleIndex).add(number);
                }
            }
        }
    }

    // 계산
    public void solve() throws IOException {
        this.calculate(0);
    }

    private boolean calculate(int n) throws IOException {
        if (n >= this.blanks.size()) {
            this.printSudoku();
            System.exit(0);
        }

        int[] currentBlank = this.blanks.get(n);
        int x = currentBlank[0];
        int y = currentBlank[1];
        int moduleIndex = (x / 3) * 3 + (y / 3);

        for (int num = 1; num < 10; num++) {
            if (
                    !this.rowMemo.get(x).contains(num) &&
                            !this.columnMemo.get(y).contains(num) &&
                            !this.moduleMemo.get(moduleIndex).contains(num)
            ) {
                this.sudoku[x][y] = num;
                this.rowMemo.get(x).add(num);
                this.columnMemo.get(y).add(num);
                this.moduleMemo.get(moduleIndex).add(num);

                if (calculate(n + 1)) return true;

                this.sudoku[x][y] = 0;
                this.rowMemo.get(x).remove(num);
                this.columnMemo.get(y).remove(num);
                this.moduleMemo.get(moduleIndex).remove(num);
            }
        }

        return false;
    }

    private void printSudoku() throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        for (int[] ints : this.sudoku) {
            for (int anInt : ints) {
                bw.write(anInt + " ");
            }
            bw.newLine();
        }

        bw.flush();
        bw.close();
    }
}
